messaging/src/main/java/io/lantern/messaging/tassis/Client.kt

```diff
@@ -0,0 +1,474 @@
+/*
+ * This Kotlin source file was generated by the Gradle 'init' task.
+ */
+package io.lantern.messaging.tassis
+
+import com.google.protobuf.ByteString
+import mu.KotlinLogging
+import org.whispersystems.libsignal.DeviceId
+import org.whispersystems.libsignal.SignalProtocolAddress
+import org.whispersystems.libsignal.ecc.ECPublicKey
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+
+internal val logger = KotlinLogging.logger {}
+
+class CloseTimedOutException : Exception("Close timed out")
+
+/**
+ * Callback interface for receiving asynchronous results.
+ */
+interface Callback<T> {
+    fun onSuccess(result: T)
+
+    fun onError(err: Throwable)
+}
+
+internal class AckCallback(private val cb: Callback<Unit>) : Callback<Messages.Ack> {
+    override fun onSuccess(result: Messages.Ack) {
+        try {
+            cb.onSuccess(Unit)
+        } catch (t: Throwable) {
+            logger.error("error on client callback: $t.message", t)
+        }
+    }
+
+    override fun onError(err: Throwable) {
+        cb.onError(err)
+    }
+}
+
+/**
+ * A handler for messages received from a transport (e.g. a websocket connection)
+ */
+interface MessageHandler {
+    /**
+     * A place for the Transport to register itself
+     */
+    fun setTransport(transport: Transport)
+
+    /**
+     * Called if there's an error connecting to or communicating with the Transport
+     */
+    fun onFailure(err: Throwable)
+
+    /**
+     * Called whenever a new message arrives from the remote end
+     */
+    fun onMessage(data: ByteArray?)
+
+    /**
+     * Called when the transport has been closed to let the MessageHandler know that it should close
+     * too.
+     */
+    fun onClose()
+}
+
+/**
+ * A facility for sending messages (e.g. a websocket connection)
+ */
+interface Transport {
+    fun send(data: ByteArray)
+
+    /**
+     * Immediately closes the transport
+     */
+    fun cancel()
+
+    /**
+     * Initiates a graceful close of the transport
+     */
+    fun close()
+}
+
+/**
+ * A factory for Transports.
+ */
+interface TransportFactory {
+    fun connect(handler: MessageHandler)
+}
+
+/**
+ * Represents an error from tassis
+ */
+data class TassisError(val name: String, val description: String) :
+    RuntimeException("$name: $description")
+
+/**
+ * Represents an inbound message from another user. Call ack() once the message has been durably
+ * recorded so that it can be deleted server-side.
```

Comment:
We don't yet support multiple devices, but once we do it will work because messages are actually delivered to devices, not just users. So when the client connects to tassis, it's connecting as a specific user and a specific device. When the sender sends the message, they're sending it to a specific user and device. If the user has multiple devices, it's the sender's responsibility to send the message to each device separately.

Importantly, each device maintains its own session keys, so it's not possible to encrypt a message just once for consumption on multiple devices. That's just how Signal works, not a choice of our own.

Reply:
Interesting, I see.  As a user, I would prefer messages to be seamlessly synced on all of my devices.  However, I do understand that there might be people with more sensitive situations who would not prefer that.  It's nice that we can support their use case.

FWIW, the default for Signal seems to be syncing messages between all devices.  It sounds like it's possible we could offer the same, either as an opt-in or opt-out setting.

Reply:
That's definitely the plan when we add multi device support. Signal actually does the syncing by having your devices communicate with each other using signal messages. They still have their own separate keys.